File: ml/action_space.py
"""
Action Space Mapping for Tetris AlphaZero

Maps between action indices (0-733) and (x, y, rotation) placement tuples.
The action space covers all positions that are valid for at least one piece
on an empty board.
"""

import numpy as np
from typing import Optional

# Tetromino shapes: [piece_type][rotation][row][col]
# Same as in Rust: I, O, T, S, Z, J, L
TETROMINOS = [
    # I piece - index 0
    [
        [[0,0,0,0], [1,1,1,1], [0,0,0,0], [0,0,0,0]],  # rot 0
        [[0,0,1,0], [0,0,1,0], [0,0,1,0], [0,0,1,0]],  # rot 1
        [[0,0,0,0], [0,0,0,0], [1,1,1,1], [0,0,0,0]],  # rot 2
        [[0,1,0,0], [0,1,0,0], [0,1,0,0], [0,1,0,0]],  # rot 3
    ],
    # O piece - index 1
    [
        [[0,0,0,0], [0,1,1,0], [0,1,1,0], [0,0,0,0]],
        [[0,0,0,0], [0,1,1,0], [0,1,1,0], [0,0,0,0]],
        [[0,0,0,0], [0,1,1,0], [0,1,1,0], [0,0,0,0]],
        [[0,0,0,0], [0,1,1,0], [0,1,1,0], [0,0,0,0]],
    ],
    # T piece - index 2
    [
        [[0,1,0,0], [1,1,1,0], [0,0,0,0], [0,0,0,0]],
        [[0,1,0,0], [0,1,1,0], [0,1,0,0], [0,0,0,0]],
        [[0,0,0,0], [1,1,1,0], [0,1,0,0], [0,0,0,0]],
        [[0,1,0,0], [1,1,0,0], [0,1,0,0], [0,0,0,0]],
    ],
    # S piece - index 3
    [
        [[0,1,1,0], [1,1,0,0], [0,0,0,0], [0,0,0,0]],
        [[0,1,0,0], [0,1,1,0], [0,0,1,0], [0,0,0,0]],
        [[0,0,0,0], [0,1,1,0], [1,1,0,0], [0,0,0,0]],
        [[1,0,0,0], [1,1,0,0], [0,1,0,0], [0,0,0,0]],
    ],
    # Z piece - index 4
    [
        [[1,1,0,0], [0,1,1,0], [0,0,0,0], [0,0,0,0]],
        [[0,0,1,0], [0,1,1,0], [0,1,0,0], [0,0,0,0]],
        [[0,0,0,0], [1,1,0,0], [0,1,1,0], [0,0,0,0]],
        [[0,1,0,0], [1,1,0,0], [1,0,0,0], [0,0,0,0]],
    ],
    # J piece - index 5
    [
        [[1,0,0,0], [1,1,1,0], [0,0,0,0], [0,0,0,0]],
        [[0,1,1,0], [0,1,0,0], [0,1,0,0], [0,0,0,0]],
        [[0,0,0,0], [1,1,1,0], [0,0,1,0], [0,0,0,0]],
        [[0,1,0,0], [0,1,0,0], [1,1,0,0], [0,0,0,0]],
    ],
    # L piece - index 6
    [
        [[0,0,1,0], [1,1,1,0], [0,0,0,0], [0,0,0,0]],
        [[0,1,0,0], [0,1,0,0], [0,1,1,0], [0,0,0,0]],
        [[0,0,0,0], [1,1,1,0], [1,0,0,0], [0,0,0,0]],
        [[1,1,0,0], [0,1,0,0], [0,1,0,0], [0,0,0,0]],
    ],
]

PIECE_NAMES = ["I", "O", "T", "S", "Z", "J", "L"]
NUM_PIECE_TYPES = 7
BOARD_WIDTH = 10
BOARD_HEIGHT = 20


def get_cells(piece_type: int, rotation: int, x: int, y: int) -> list[tuple[int, int]]:
    """Get the cells occupied by a piece at position (x, y) with given rotation."""
    shape = TETROMINOS[piece_type][rotation]
    cells = []
    for dy in range(4):
        for dx in range(4):
            if shape[dy][dx] == 1:
                cells.append((x + dx, y + dy))
    return cells


def is_valid_position(piece_type: int, rotation: int, x: int, y: int) -> bool:
    """Check if a piece fits on an empty board at this position."""
    cells = get_cells(piece_type, rotation, x, y)
    for cx, cy in cells:
        if cx < 0 or cx >= BOARD_WIDTH:
            return False
        if cy < 0 or cy >= BOARD_HEIGHT:
            return False
    return True


def _build_action_mappings() -> tuple[list[tuple[int, int, int]], dict[tuple[int, int, int], int]]:
    """Build the action index mappings.

    Returns:
        ACTION_TO_PLACEMENT: List where index i gives (x, y, rotation) tuple
        PLACEMENT_TO_ACTION: Dict mapping (x, y, rotation) -> action index
    """
    # Search beyond 0-9, 0-19 because pieces have offsets within their 4x4 grid
    X_MIN, X_MAX = -3, BOARD_WIDTH
    Y_MIN, Y_MAX = -3, BOARD_HEIGHT

    # Track all positions valid for ANY piece
    valid_positions = set()

    for y in range(Y_MIN, Y_MAX):
        for x in range(X_MIN, X_MAX):
            for rot in range(4):
                for piece_type in range(NUM_PIECE_TYPES):
                    if is_valid_position(piece_type, rot, x, y):
                        valid_positions.add((x, y, rot))
                        break  # Position is valid if any piece fits

    # Sort for deterministic ordering: by rotation, then y, then x
    sorted_positions = sorted(valid_positions, key=lambda p: (p[2], p[1], p[0]))

    action_to_placement = sorted_positions
    placement_to_action = {pos: idx for idx, pos in enumerate(sorted_positions)}

    return action_to_placement, placement_to_action


# Build the mappings at module load time
ACTION_TO_PLACEMENT, PLACEMENT_TO_ACTION = _build_action_mappings()
NUM_ACTIONS = len(ACTION_TO_PLACEMENT)


def action_to_placement(action_idx: int) -> tuple[int, int, int]:
    """Convert action index to (x, y, rotation) placement tuple."""
    if action_idx < 0 or action_idx >= NUM_ACTIONS:
        raise ValueError(f"Invalid action index: {action_idx}. Must be in [0, {NUM_ACTIONS})")
    return ACTION_TO_PLACEMENT[action_idx]


def placement_to_action(x: int, y: int, rotation: int) -> Optional[int]:
    """Convert (x, y, rotation) to action index, or None if invalid."""
    return PLACEMENT_TO_ACTION.get((x, y, rotation))


def get_action_mask(board: np.ndarray, piece_type: int) -> np.ndarray:
    """
    Generate a binary mask of valid actions for the given piece and board state.

    Args:
        board: 2D numpy array of shape (20, 10), 1 = filled, 0 = empty
        piece_type: Integer 0-6 representing piece type

    Returns:
        1D numpy array of shape (NUM_ACTIONS,), 1 = valid, 0 = invalid
    """
    mask = np.zeros(NUM_ACTIONS, dtype=np.float32)

    for action_idx, (x, y, rot) in enumerate(ACTION_TO_PLACEMENT):
        if _piece_can_be_placed(board, piece_type, x, y, rot):
            mask[action_idx] = 1.0

    return mask


def _piece_can_be_placed(board: np.ndarray, piece_type: int, x: int, y: int, rotation: int) -> bool:
    """Check if a piece can be placed at the given position on the board."""
    cells = get_cells(piece_type, rotation, x, y)

    for cx, cy in cells:
        # Check bounds
        if cx < 0 or cx >= BOARD_WIDTH:
            return False
        if cy < 0 or cy >= BOARD_HEIGHT:
            return False
        # Check collision with existing blocks
        if board[cy, cx] != 0:
            return False

    return True


def get_valid_actions(board: np.ndarray, piece_type: int) -> list[int]:
    """
    Get list of valid action indices for the given piece and board state.

    Args:
        board: 2D numpy array of shape (20, 10), 1 = filled, 0 = empty
        piece_type: Integer 0-6 representing piece type

    Returns:
        List of valid action indices
    """
    valid = []
    for action_idx, (x, y, rot) in enumerate(ACTION_TO_PLACEMENT):
        if _piece_can_be_placed(board, piece_type, x, y, rot):
            valid.append(action_idx)
    return valid


# Pre-compute per-piece valid actions on empty board for quick reference
_PIECE_VALID_ACTIONS_EMPTY = {}
for _pt in range(NUM_PIECE_TYPES):
    _empty_board = np.zeros((BOARD_HEIGHT, BOARD_WIDTH), dtype=np.uint8)
    _PIECE_VALID_ACTIONS_EMPTY[_pt] = get_valid_actions(_empty_board, _pt)


def get_piece_valid_actions_empty(piece_type: int) -> list[int]:
    """Get valid action indices for a piece on an empty board (cached)."""
    return _PIECE_VALID_ACTIONS_EMPTY[piece_type]


if __name__ == "__main__":
    # Print statistics
    print(f"Total action space size: {NUM_ACTIONS}")
    print()

    # Count by rotation
    by_rot = [0, 0, 0, 0]
    for x, y, rot in ACTION_TO_PLACEMENT:
        by_rot[rot] += 1
    print(f"Actions by rotation: {by_rot}")
    print()

    # Show range of x, y values
    xs = [p[0] for p in ACTION_TO_PLACEMENT]
    ys = [p[1] for p in ACTION_TO_PLACEMENT]
    print(f"X range: [{min(xs)}, {max(xs)}]")
    print(f"Y range: [{min(ys)}, {max(ys)}]")
    print()

    # Valid actions per piece on empty board
    print("Valid actions per piece (empty board):")
    for piece_type in range(NUM_PIECE_TYPES):
        count = len(get_piece_valid_actions_empty(piece_type))
        print(f"  {PIECE_NAMES[piece_type]}: {count}")

File: tetris_core/src/mcts/action_space.rs
//! Action Space for Tetris MCTS
//!
//! Maps (x, y, rotation) placements to action indices 0-733.

use std::collections::HashMap;
use std::sync::OnceLock;

use crate::constants::{BOARD_HEIGHT, BOARD_WIDTH};
use crate::piece::NUM_PIECE_TYPES;

/// Global cached ActionSpace (initialized once on first use)
static ACTION_SPACE: OnceLock<ActionSpace> = OnceLock::new();

/// Number of actions in the action space
pub const NUM_ACTIONS: usize = 734;

/// Action index mapping
/// Maps (x, y, rotation) to action index 0-733
/// Built at module load time to match Python's action_space.py
#[derive(Clone)]
pub struct ActionSpace {
    pub action_to_placement: Vec<(i32, i32, usize)>, // (x, y, rotation)
    pub placement_to_action: HashMap<(i32, i32, usize), usize>,
}

impl ActionSpace {
    pub fn new() -> Self {
        let mut action_to_placement = Vec::new();
        let mut placement_to_action = HashMap::new();

        // Same logic as Python's action_space.py
        let x_min = -3i32;
        let x_max = 10i32;
        let y_min = -3i32;
        let y_max = 20i32;

        // Check which positions are valid for at least one piece
        let mut valid_positions: Vec<(i32, i32, usize)> = Vec::new();

        for y in y_min..y_max {
            for x in x_min..x_max {
                for rot in 0..4 {
                    // Check if any piece fits at this position on an empty board
                    for piece_type in 0..NUM_PIECE_TYPES {
                        if Self::is_valid_position_empty_board(piece_type, rot, x, y) {
                            valid_positions.push((x, y, rot));
                            break;
                        }
                    }
                }
            }
        }

        // Sort by rotation, then y, then x (to match Python)
        valid_positions.sort_by_key(|&(x, y, rot)| (rot, y, x));

        for (idx, pos) in valid_positions.iter().enumerate() {
            action_to_placement.push(*pos);
            placement_to_action.insert(*pos, idx);
        }

        ActionSpace {
            action_to_placement,
            placement_to_action,
        }
    }

    fn is_valid_position_empty_board(piece_type: usize, rotation: usize, x: i32, y: i32) -> bool {
        let shape = &crate::piece::TETROMINOS[piece_type][rotation];
        for dy in 0..4 {
            for dx in 0..4 {
                if shape[dy][dx] == 1 {
                    let cx = x + dx as i32;
                    let cy = y + dy as i32;
                    if cx < 0 || cx >= BOARD_WIDTH as i32 || cy < 0 || cy >= BOARD_HEIGHT as i32 {
                        return false;
                    }
                }
            }
        }
        true
    }

    pub fn num_actions(&self) -> usize {
        self.action_to_placement.len()
    }

    pub fn placement_to_index(&self, x: i32, y: i32, rotation: usize) -> Option<usize> {
        self.placement_to_action.get(&(x, y, rotation)).copied()
    }

    pub fn index_to_placement(&self, idx: usize) -> Option<(i32, i32, usize)> {
        self.action_to_placement.get(idx).copied()
    }
}

impl Default for ActionSpace {
    fn default() -> Self {
        ActionSpace::new()
    }
}

/// Get the global cached action space (initialized once on first use)
pub fn get_action_space() -> &'static ActionSpace {
    ACTION_SPACE.get_or_init(ActionSpace::new)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_action_space() {
        let action_space = ActionSpace::new();
        assert_eq!(action_space.num_actions(), 734);

        // Test roundtrip
        for (idx, &(x, y, rot)) in action_space.action_to_placement.iter().enumerate() {
            let idx2 = action_space.placement_to_index(x, y, rot).unwrap();
            assert_eq!(idx, idx2);
        }
    }
}
